### PR TITLE
Improve Bond's compliance with Microsoft's SDL

### PR DESCRIPTION
**Pull request by chwarr**
For projects that use Microsoft's SDL, these commits help minimize the amount of work needed to use Bond.

As a bonus, the changes are relatively contained, and don't materially affect Clang or GCC.

These commits should be rebased atop master, or a full merge should be done.

### CODE REVIEW

**Comment by chwarr** [comment]
I've broken something somewhere. I'm taking a look.

**Comment by chwarr** [comment]
Fixes for Clang & GCC and VC++ 2013 have been integrated. They've been rebased in, but no one has yet reviewed and left comments, so I force pushed. If you want to see older revisions, let me know the commit hash, and I'll resurrect it.

File: examples/cpp/core/multiprecision/CMakeLists.txt [hunk 1]
```diff
@@ -8,3 +8,7 @@ add_bond_codegen (multiprecision.bond
 add_bond_test (multiprecision
     ${CMAKE_CURRENT_BINARY_DIR}/${CMAKE_CFG_INTDIR}/multiprecision_types.cpp
     multiprecision.cpp)
+
+# Deep in the guts of Boost we get a warning about not using a "secure"
+# routine. Suppress them for this example.
+cxx_target_compile_options(MSVC multiprecision PRIVATE -D_SCL_SECURE_NO_WARNINGS)
```
Comment (chadwalters) [inline examples/cpp/core/multiprecision/CMakeLists.txt]: We may need to track this down at some point if Bond is pulling this in for our customers. If there is some specific behavior in this example that pulls it in unrelated to the core of Bond then I am less concerned about it.

File: examples/cpp/core/bf/err.h [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+// Convert an error number into a human-readable string.
+inline std::string ErrorString(int errnum)
+{
+    std::string result;
+
+#ifdef _MSC_VER
+    // strerrorlen_s hasn't made it into the Microsoft C runtime yet. Until
+    // it has, we're just going to reserve a buffer that's pretty big and
+    // get as much of the error message as we can.
+    result.reserve(240);
```
Comment (tstein) [inline examples/cpp/core/bf/err.h]: I feel like you can make this bigger if you want to.

File: examples/cpp/core/multiprecision/CMakeLists.txt (hunk 1 above)
Comment (chadwalters) [inline examples/cpp/core/multiprecision/CMakeLists.txt]: We discussed and this doesn't seem like a big concern.

**Comment by chadwalters** [comment]
CI failure on VC14 build is not related to this change. Looks good to go.

File: examples/cpp/core/multiprecision/CMakeLists.txt (hunk 1 above)
Comment (chwarr) [inline examples/cpp/core/multiprecision/CMakeLists.txt]: Yeah, the Boost stuff emitting a warning is being pulled in by the example in question, not by Bond.

**Pull request merged**